core/templates/pages/exploration-player-page/services/question-player-engine.service.ts

```diff
@@ -16,285 +16,315 @@
-        .loadExploration(_explorationId, version)
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+import { AppConstants } from 'app.constants';
+import { ReadOnlyExplorationBackendApiService } from 'domain/exploration/read-only-exploration-backend-api.service';
+import { BindableVoiceovers } from 'domain/exploration/RecordedVoiceoversObjectFactory';
+import { Question, QuestionBackendDict, QuestionObjectFactory } from 'domain/question/QuestionObjectFactory';
+import { State } from 'domain/state/StateObjectFactory';
+import { StateCard, StateCardObjectFactory } from 'domain/state_card/StateCardObjectFactory';
+import { ExpressionInterpolationService } from 'expressions/expression-interpolation.service';
+import { InteractionAnswer } from 'interactions/answer-defs';
+import { AnswerClassificationService, InteractionRulesService } from 'pages/exploration-player-page/services/answer-classification.service';
+import { InteractionSpecsConstants } from 'pages/interaction-specs.constants';
+import { AlertsService } from 'services/alerts.service';
+import { ContextService } from 'services/context.service';
+import { UrlService } from 'services/contextual/url.service';
+import { ExplorationHtmlFormatterService } from 'services/exploration-html-formatter.service';
+import { FocusManagerService } from 'services/stateful/focus-manager.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class QuestionPlayerEngineService {
+  private explorationId: string = null;
+  private questionPlayerMode: boolean = null;
+  private version: number = null;
+
+  private answerIsBeingProcessed: boolean = false;
+  private questions: Question[] = [];
+  private currentIndex: number = null;
+  private nextIndex: number = null;
+
+  constructor(
+      private alertsService: AlertsService,
+      private answerClassificationService: AnswerClassificationService,
+      private contextService: ContextService,
+      private explorationHtmlFormatterService: ExplorationHtmlFormatterService,
+      private expressionInterpolationService: ExpressionInterpolationService,
+      private focusManagerService: FocusManagerService,
+      private questionObjectFactory: QuestionObjectFactory,
+      private readOnlyExplorationBackendApiService:
+        ReadOnlyExplorationBackendApiService,
+      private stateCardObjectFactory: StateCardObjectFactory,
+      private urlService: UrlService) {
+    contextService.setQuestionPlayerIsOpen();
+    this.explorationId = contextService.getExplorationId();
+    this.questionPlayerMode = contextService.isInQuestionPlayerMode();
+    this.version = urlService.getExplorationVersionFromUrl();
+
+    if (!this.questionPlayerMode) {
+      readOnlyExplorationBackendApiService
+        .loadExploration(this.explorationId, this.version)
         .then(function(exploration) {
-          version = exploration.version;
-          $rootScope.$applyAsync();
```

Comment:
@kevintab95 @srijanreddy98 I am getting following E2E error: 

Expected $[5] = Entry({ level: SEVERE, message: 'webpack:///./core/templates/services/contextual/logger.service.ts? 59:16 "Please enter a valid exploration ID."', timestamp: 1607059532400, type: '' }) to equal undefined.

Could this be due to the "rootScope.ApplyAsync" call missing after my changes ? If so, how could we handle it ?

Reply:
Hi @subedi90, it appears to be the case that, in the context service, "questionPlayerIsManuallySet" is incorrectly set to be true even when the exploration player is being viewed. This causes "contextService.getExplorationId()" to return an undefined value when it is used. This is causing errors in the e2e tests. You can repro this manually by trying to play any exploration locally with the dev console open. Hope this helps!

Reply:
Thanks for the tips @kevintab95 . This certainly did help. I moved the questionPlayerMode assignment (and all other assignments) from constructor to init() function. I am exactly not sure if the object instatiation order got changed due to the migration and caused this issue, but moving those assignments to init() seems logical given those are not used before call to init().

Reply:
With this change all tests passes !

Reply:
Great to hear that this has been fixed. Thanks!